src/local/pydoit_nb/tasks_notebooks.py

```diff
@@ -0,0 +1,134 @@
+"""
+Task generation related to notebooks
+"""
+from __future__ import annotations
+
+from collections.abc import Iterable
+from pathlib import Path
+from typing import Protocol, TypeVar
+
+from .config_handling import get_branch_config_ids
+from .notebook import ConfiguredNotebook, UnconfiguredNotebook
+from .notebook_step import NotebookStep
+from .typing import ConfigBundleLike, Converter, DoitTaskSpec, HandleableConfiguration
+
+T = TypeVar("T")
+
+
+class GetUnconfiguredNotebooksCallable(Protocol):
+    """Callable that can be used for getting unconfigured notebooks"""
+
+    def __call__(self) -> list[UnconfiguredNotebook]:  # noqa: D102
+        ...  # pragma: no cover
+
+
+class ConfigureNotebooksCallable(Protocol[T]):
+    """Callabale that can be used for configuring notebooks"""
+
+    def __call__(  # noqa: D102
+        self,
+        unconfigured_notebooks: Iterable[UnconfiguredNotebook],
+        config_bundle: ConfigBundleLike[T],
+        branch_name: str,
+        branch_config_id: str,
+    ) -> list[ConfiguredNotebook]:
+        ...  # pragma: no cover
+
+
+def get_notebook_branch_tasks(  # noqa: PLR0913
+    branch_name: str,
+    get_unconfigured_notebooks: GetUnconfiguredNotebooksCallable,
+    configure_notebooks: ConfigureNotebooksCallable[T],
+    config_bundle: ConfigBundleLike[T],
+    root_dir_raw_notebooks: Path,
+    converter: Converter[tuple[HandleableConfiguration, ...]] | None = None,
+    clean: bool = True,
+) -> Iterable[DoitTaskSpec]:
+    """
+    Get tasks for the notebooks within a given notebok branch
+
+    A notebook branch is a group of notebooks within the wider workflow. These
```

Comment:
I don't quite understand this section. So in this case each branch would be a folder in `notebooks`?  Or is conceptually more similar to our historical - scenario branching in the CSIRO project? 

Obviously the docs would cover this in due course, but I had to dig to figure out what you meant by branches

Reply:
I think I'll have to draw it to explain it well. Let's do that on zoom sometime then I can try and capture here as best as I can

Reply:
Based on conversation from today, I'll update to `UnconfiguredStep` and we can see how it looks after that. Docs will be something like:

A collection of unconfigured notebooks which form a step in the workflow. How you group notebooks to form these steps doesn't actually affect the wider workflow (notebooks are converted to doit tasks at the individual level and doit handles forming a directed acyclic graph based on their targets and dependencies) but it can be helpful for simplifying the links between notebooks and for understanding what depends on what and how. This explanation is not our best so if you would like more help or have a suggestion for a better explanation please raise an issue at [TODO link to relevant place for issues, probably pydoit-nb].

For our records, the diagram I drew today: [Climate Resource Meeting 2023-11-22 08_49 AM-[1700639986226].pdf](https://github.com/climate-resource/CMIP-GHG-Concentration-Generation/files/13435778/Climate.Resource.Meeting.2023-11-22.08_49.AM-.1700639986226.pdf)

Reply:
Fully refactored. Add comments in new threads as needed.